src/Hash.hs

```diff
@@ -44,31 +43,31 @@ inlineHashes hashedPaths bucketFiles@BucketFiles{..} =
     prependWithHash :: MonadThrow m => Paths -> Text -> Text -> m Text
     prependWithHash hashes path altPath =
       maybe (throwHashNotFound path)
-            (pure . pack . Posix.joinPath . (: [unpack altPath]) . unpack)
+            (pure . pack . joinPath . (: [unpack altPath]) . unpack)
             (HashMap.lookup path hashes)
       where
         throwHashNotFound filePath = throwM $ HashNotFound filePath
 
-hashBucketFiles :: (MonadIO m, MonadThrow m) => BucketFiles -> m FileHashes
+hashBucketFiles :: (MonadThrow m, MonadBaseControl IO m, MonadIO m) => BucketFiles -> m FileHashes
 hashBucketFiles BucketFiles{..} =
   if _isHashed
     then HashMap.traverseWithKey hashKey _paths
     else return HashMap.empty
   where
-    hashKey :: (MonadIO m, MonadThrow m) => Text -> a -> m Text
+    hashKey :: (MonadThrow m, MonadBaseControl IO m, MonadIO m) => Text -> a -> m Text
     hashKey path _ = do
       hash <- createHash $ unpack path
       return $ (pack . show) hash
 
-createHash :: (MonadThrow m, MonadIO m) => FilePath -> m (Digest SHA1)
-createHash path = do
-  files <- (liftIO . doesDirectoryExist) path >>= \case
-    True -> do
-      directoryFiles <- (fmap . fmap) (\x -> [path, x]) (liftIO . listDirectory $ path)
-      return $ fmap joinPath directoryFiles
+createHash :: (MonadThrow m, MonadBaseControl IO m, MonadIO m)
+                    => FilePath
+                    -> m (Digest SHA1)
+createHash path =
+  (liftIO . doesDirectoryExist) path >>= \case
```

Comment:
I would wrap both `doesDirectoryExist` and `doesFileExist` into a `checkDirectory :: FilePath -> IO (Maybe FilePath)` and `checkFile` along the lines of https://www.stackage.org/haddock/lts-7.13/base-4.9.0.0/Data-List.html#v:stripPrefix, so you can then use PatternGuards to make a single total pattern match for these cases. Happy to explain further if this doesn't make sense.